File: leetcode3/변지협/1905. Count Sub Islands.py
'''
1. 아이디어 :
    grid2 bfs를 2중 for문 돌고, 
    각 bfs 돌면서 grid2 (인접한 곳) 을 visited 에 저장한다.
    visited에 저장된 곳들을 grid1과 비교해서 sub island인지 판단하고, 모두 0으로 초기화한다.
2. 시간복잡도 :
    O(n*m)
3. 자료구조/알고리즘 :
    bfs
'''


class Solution:
    def countSubIslands(self, grid1, grid2):
        directions = [(-1,0),(1,0),(0,-1),(0,1)]
        answer = 0

        y_len = len(grid2)
        x_len = len(grid2[0])
        
        for y in range(y_len):
            for x in range(x_len):
                queue = deque()
                visited = set()
                if grid2[y][x] == 1:
                    queue.append((x,y))
                    visited.add((x,y))
                
                while True:
                    if len(queue) == 0:
                        break
                    
                    _x,_y = queue.popleft()
                    
                    for dx,dy in directions:
                        if 0 <= dx + _x < x_len and 0 <= dy + _y < y_len and \
                         grid2[_y+dy][_x+dx] == 1 and (_x+dx, _y+dy) not in visited:
                            queue.append((_x+dx,_y+dy))
                            visited.add((_x+dx,_y+dy))
                
                sub = True
                for _x,_y in visited:
                    if grid1[_y][_x] == 0:
                        sub = False
                    grid2[_y][_x] = 0
                
                if visited and sub:
                    answer +=1
        
        return answer
                # print(visited)
                        
                

File: leetcode3/이진희/1905_Count Sub Islands.java
/*

1. 아이디어 : 인접행렬 그래프 탐색 문제
            grid2 격자 기준으로 섬을 탐색한다, 연결된 섬을 탐색하는 과정에서 grid1의 같은 위치에 하나라도 바다가 있다면, sub island를 만들 수 없으므로 false 반환
            전부 grid1 섬과 일치한다면, 포함 가능하다는 것이므로 true를 반환해준다

2. 시간복잡도 : O(N*M)

3. 자료구조/알고리즘 : bfs

 */

class Solution {
    private int lenY, lenX;
    private int[] dy = {0,1,0,-1};
    private int[] dx = {-1,0,1,0};
    private boolean[][] visited;

    // 0 = 바다
    // 1 = 섬
    public int countSubIslands(int[][] grid1, int[][] grid2) {
        lenY = grid1.length;
        lenX = grid1[0].length;

        visited = new boolean[lenY][lenX];
        int ans = 0;

        for(int i=0; i<lenY; i++) {
            for(int j=0; j<lenX; j++) {
                if(visited[i][j] || grid2[i][j] == 0) continue;
                if(bfs(i,j,grid1,grid2)) ans++;
            }
        }

        return ans;
    }

    private boolean bfs(int startY, int startX, int[][] grid1, int[][] grid2) {
        Deque<int[]> dq = new ArrayDeque<>();
        boolean check = true;

        dq.add(new int[]{startY, startX});
        visited[startY][startX] = true;

        while(!dq.isEmpty()) {
            int[] curr = dq.poll();
            int y = curr[0];
            int x = curr[1];

            if(check && grid1[y][x] == 0) check = false;

            for(int dir=0; dir<4; dir++) {
                int ny = y + dy[dir];
                int nx = x + dx[dir];

                if(ny<0 || nx<0 || ny>=lenY || nx>=lenX) continue;
                if(visited[ny][nx] || grid2[ny][nx] == 0) continue;

                dq.add(new int[]{ny,nx});
                visited[ny][nx] = true;
            }
        }

        return check;
    }
}